From eb259b308bd6225c7d9bda5cc3f4a89c88c5f5d1 Mon Sep 17 00:00:00 2001
From: Luca Stefani <luca.stefani.ge1@gmail.com>
Date: Tue, 6 Nov 2018 15:08:45 +0100
Subject: [PATCH] [SQUSH][DNM] Merge tag 'android-9.0.0_r16' into lineage-16.0

commit f172c8512a205b8403eebfdabc2cfefe82afc772
Merge: c514b08 7cc2085
Author: android-build-team Robot <android-build-team-robot@google.com>
Date:   Fri Aug 24 22:00:54 2018 +0000

    Merge cherrypicks of [4871059, 4868580, 4869372] into pi-qpr1-release

    Change-Id: I9c8f77bf25e97f0f906ea31474f3b492d1dd3a04

commit 7cc2085e7ee20c950a6ad37a9619dc8216ff75e6
Author: Eva Bertels <evabertels@google.com>
Date:   Thu Aug 16 12:58:39 2018 +0100

    Add check for misprovisioned Pixel 2 device.

    Some Pixel devices had a wrong brand value provisioned into keymaster.
    Due to this misprovisioning those devices fail device ID attestation because it includes a check for the correct brand value.
    This is now solved by re-trying Device ID attestation if we are running
    on a potentially misprovisioned device, allowing for the known incorrect
    brand value.

    Bug: 69471841
    Test: atest com.android.cts.devicepolicy.MixedDeviceOwnerTest#testKeyManagement
    Change-Id: I99108659f9a7b65d4f80f0a4631a382bfb076738
    Merged-In: I6737184eb5a34cf3213f9cb1c5d5e2f1cf02ea38
    (cherry picked from commit 74bf12f6b564fb136ab19532dad63ae985350060)

commit c514b08c1074c95d72acac64759f8e614b4d1d85
Author: Bill Yi <byi@google.com>
Date:   Wed Jul 4 18:26:58 2018 -0700

    Import translations. DO NOT MERGE

    Auto-generated-cl: translation import

    Bug: 64712476
    Change-Id: I87c57ae6137e607e2d147e31758eca5e7cd3f1b9

commit 7ea01865d5bb00f4911134941c5bd521c48fd5a6
Author: Bill Yi <byi@google.com>
Date:   Thu Jun 21 10:12:40 2018 -0700

    Import translations. DO NOT MERGE

    Auto-generated-cl: translation import

    Bug: 64712476
    Change-Id: I22fc3e6bceb4df1bf5f8c36bdd0d4b7a98fd0573

Change-Id: Ib23834390f55f33e7907c29801ee155d28225b93
---
 res/values-bs/strings.xml                     |  2 +-
 res/values-hi/strings.xml                     |  2 +-
 src/com/android/keychain/KeyChainService.java | 27 ++++++++++++++++++-
 3 files changed, 28 insertions(+), 3 deletions(-)

diff --git a/res/values-bs/strings.xml b/res/values-bs/strings.xml
index 4fa2989..703ab22 100644
--- a/res/values-bs/strings.xml
+++ b/res/values-bs/strings.xml
@@ -20,7 +20,7 @@
     <string name="title_no_certs" msgid="8350009443064722873">"Certifikat nije pronađen."</string>
     <string name="title_select_cert" msgid="3588447616418041699">"Odaberite certifikat"</string>
     <string name="requesting_application" msgid="1589142627467598421">"Aplikacija %s je zatražila certifikat. Izbor certifikata će omogućiti aplikaciji korištenje ovog identiteta na serverima, sada i ubuduće."</string>
-    <string name="requesting_server" msgid="5832565605998634370">"Aplikacija je identifikovala server koji traži certifikat kao %s. Dozvolite aplikaciji da pristupi certifikatu samo ako joj vjerujete."</string>
+    <string name="requesting_server" msgid="5832565605998634370">"Aplikacija je identifikovala server koji traži certifikat kao %s. family link da pristupi certifikatu samo ako joj vjerujete."</string>
     <string name="install_new_cert_message" msgid="4451971501142085495">"Možete instalirati certifikate iz PKCS#12 fajla sa ekstenzijom %1$s ili %2$s, koji se nalazi u vanjskoj pohrani."</string>
     <string name="install_new_cert_button_label" msgid="903474285774077171">"Instaliraj certifikat"</string>
     <string name="allow_button" msgid="3030990695030371561">"Odaberi"</string>
diff --git a/res/values-hi/strings.xml b/res/values-hi/strings.xml
index d7bfde4..e49f72e 100644
--- a/res/values-hi/strings.xml
+++ b/res/values-hi/strings.xml
@@ -24,5 +24,5 @@
     <string name="install_new_cert_message" msgid="4451971501142085495">"आप बाहरी मेमोरी में मौजूद %1$s या %2$s एक्‍सटेंशन वाली PKCS#12 फ़ाइल से प्रमाणपत्र इंस्‍टॉल कर सकते हैं."</string>
     <string name="install_new_cert_button_label" msgid="903474285774077171">"प्रमाणपत्र इंस्‍टॉल करें"</string>
     <string name="allow_button" msgid="3030990695030371561">"चुनें"</string>
-    <string name="deny_button" msgid="3766539809121892584">"अस्वीकार करें"</string>
+    <string name="deny_button" msgid="3766539809121892584">"नामंज़ूर करें"</string>
 </resources>
diff --git a/src/com/android/keychain/KeyChainService.java b/src/com/android/keychain/KeyChainService.java
index e109d6b..3f17636 100644
--- a/src/com/android/keychain/KeyChainService.java
+++ b/src/com/android/keychain/KeyChainService.java
@@ -182,7 +182,7 @@ public class KeyChainService extends IntentService {
                 return KeyChain.KEY_ATTESTATION_MISSING_CHALLENGE;
             }
 
-            final KeymasterArguments attestArgs;
+            KeymasterArguments attestArgs;
             try {
                 attestArgs = AttestationUtils.prepareAttestationArguments(
                         mContext, idAttestationFlags, attestationChallenge);
@@ -190,6 +190,31 @@ public class KeyChainService extends IntentService {
                 Log.e(TAG, "Failed collecting attestation data", e);
                 return KeyChain.KEY_ATTESTATION_CANNOT_COLLECT_DATA;
             }
+            int errorCode = checkKeyChainStatus(alias, attestationChain, attestArgs);
+            if (errorCode == KeyChain.KEY_ATTESTATION_CANNOT_ATTEST_IDS) {
+                // b/69471841: id attestation might fail due to incorrect provisioning of device
+                try {
+                    attestArgs =
+                            AttestationUtils.prepareAttestationArgumentsIfMisprovisioned(
+                            mContext, idAttestationFlags, attestationChallenge);
+                    if (attestArgs == null) {
+                        return errorCode;
+                    }
+                } catch (DeviceIdAttestationException e) {
+                    Log.e(TAG, "Failed collecting attestation data "
+                            + "during second attempt on misprovisioned device", e);
+                    return KeyChain.KEY_ATTESTATION_CANNOT_COLLECT_DATA;
+                }
+            }
+
+            return checkKeyChainStatus(alias, attestationChain, attestArgs);
+        }
+
+        private int checkKeyChainStatus(
+                String alias,
+                KeymasterCertificateChain attestationChain,
+                KeymasterArguments attestArgs) {
+
             final String keystoreAlias = Credentials.USER_PRIVATE_KEY + alias;
             final int errorCode = mKeyStore.attestKey(keystoreAlias, attestArgs, attestationChain);
             if (errorCode != KeyStore.NO_ERROR) {
-- 
2.17.1

